packages/@lwc/engine-dom/src/apis/build-custom-element-constructor.ts

```diff
@@ -59,26 +62,30 @@ export function buildCustomElementConstructor(Ctor: ComponentConstructor): HTMLE
     return class extends HTMLElement {
         constructor() {
             super();
-
-            if (this.isConnected) {
-                // this if block is hit when there's already an un-upgraded element in the DOM with the same tag name.
-                hydrateComponent(this, Ctor, {});
-                hydratedCustomElements.add(this);
-            } else {
-                createVM(this, Ctor, renderer, {
-                    mode: 'open',
-                    owner: null,
-                    tagName: this.tagName,
-                });
+            if (!isNull(this.shadowRoot)) {
+                if (process.env.NODE_ENV !== 'production') {
+                    // eslint-disable-next-line no-console
+                    console.warn(
+                        `Found an existing shadow root for the custom element ${Ctor.name}. Call \`hydrateComponent\` instead.`
+                    );
+                }
+                clearNode(this.shadowRoot);
             }
+            if (this.childNodes.length > 0) {
+                if (process.env.NODE_ENV !== 'production') {
+                    // eslint-disable-next-line no-console
+                    console.warn(`Top level elements cannot have children.`);
+                }
+                clearNode(this);
```

Comment:
On second thought, I'm not sure if this is the behavior we want. Today, if you slot something into a custom element, it logs errors but [works fine](https://stackblitz.com/edit/salesforce-lwc-hobroi?file=index.html):

![Screenshot 2023-06-26 at 2 20 58 PM](https://github.com/salesforce/lwc/assets/283842/b008cf09-51b9-416a-9eac-d0fbd5885579)

With this change, someone with code like:

```html
<x-foo>
  <x-bar>
    <x-baz></x-baz>
  </x-bar>
</x-foo>
```

...will end up with an empty `<x-foo>` instead.

I feel like we may want to treat childNodes differently from the shadowRoot to support slotting. We should also check how folks are using `CustomElementConstructor` today to see if this is a concern.

Reply:
I took a quick glance at Trailhead, and they seem to only have empty components. So maybe this is fine.

Reply:
The reason I'm doing this is because this is in-line with what `createElement` does.

Also, this example doesn't work for light DOM components, they are not slotted properly.

Reply:
OK, let's go for it. Maybe we can revise this later if we get pushback (e.g. add a flag).